### DISCUSSION TITLE
Stable Window Placement (Position and Screen)

### DISCUSSION THREAD

**Discussion by halostatue** [wezterm/wezterm#6041] [Q&A]
### What Operating System(s) are you running on?

macOS

### WezTerm version

20240203-110809-5046fc22

### Ask your question!

I'm looking at converting from iTerm2, but there's a few things that I can't quite figure out how to do — or even if they are possible. I *suspect* that ⅔ of what I want to do here is possible with scripting — but after three days in the weeds of WezTerm config I am drawing blanks, but the last ⅓ may require new features if they are to be supported.

## Stable Window Placement

The critical feature for me is having stable window placement on the same screen when Wezterm opens (and preferably when I hit ⌘N). The way I work is generally to have all terminal windows stacked on the right side of my main (laptop) screen. To that end, in iTerm2, I have the following set in `Profiles > Default > Window` settings:

- 118 columns, 65 rows (I think that wezterm is more compact than iTerm2 compact mode, so it might be 67 rows required)
- Style: Full height Right of Screen
- Screen: Prefer main screen
- Space: Current Space

The **style** options are Normal, Full Screen, Maximized (`window:gui_window():maximize()`), No Title Bar, _Full-Width Bottom of Screen_, _Full-Width Top of Screen_, _Full-Height Left of Screen_, _Full-Height Right of Screen_, _Bottom of Screen_, _Top of Screen_, _Left of Screen_, and _Right of Screen_.

I’ve seen #5501 on how to centre a window, and I think this *should* be feasible — but my current solution:

```lua
    local main_screen = wezterm.gui.screens().main
    local x = main_screen.width - 1440
    local y = 0

    cmd.position = {
      x = x,
      y = y,
      origin = "MainScreen",
    }
```

isn't working in the `gui-startup` event. Because the window hasn't been spawned, I don't think I can get the `pixel_width`, although it might be possible.

### Feature Request 1

It would be nice to add a parameter to `maximize()` to control the maximization angle.

```lua
window:gui_window():maximize() -- same as maximize("Both")
window:gui_window():maximize("Both")
window:gui_window():maximize("Width")
window:gui_window():maximize("Height")
```

### Feature Request 2

It would be nice to have *logical* window positioning primitives. Instead of positioning the top left corner of a window with `set_position(x, y)`, I’d want to do something like `gui_window:set_logical_position { top = 0, right = 0 }` to put the window `top` pixels from the top of the screen (more or less what the `y` position is now) and `right` pixels from the right of the screen (this would be akin to setting `x` to something like `screen.width - gui_window.pixel_width`).

For the case where the window hasn't been spawned yet, it would be useful to get the *predicted* size of the window somehow so that `cmd.position` can be set better.

### Feature Request 3

It would be nice to set a screen affinity at the top level. I basically never want a terminal window on my secondary screen.

**Comment by halostatue** [comment]
The features mentioned here are still of value, but I have the basic feature working. I cannot get the *dynamic* form working, only the static form.

### Static

This uses a fixed width of 1440, which is measured from an open GUI window. It does work, though, once I fixed some logical errors.

```lua
  wezterm.on("gui-startup", function(cmd)
    local main_screen = wezterm.gui.screens().main
    local position = {
      x = main_screen.width - 1440,
      y = 0,
      origin = "MainScreen",
    }

    cmd = cmd or { position = position }

    wezterm.mux.spawn_window(cmd)
  end)
```

### Dynamic

I would like to do something more dynamic, but this doesn't work:

```lua
  wezterm.on("gui-startup", function(cmd)
    local _tab, _pane, window = wezterm.mux.spawn_window(cmd or {})
    local gui_window = window:gui_window()
    local dimensions = gui_window:get_dimensions()
    local main_screen = wezterm.gui.screens().main
    local x = main_screen.width - gui_window.pixel_width
    local y = 0

    gui_window:set_position(x, y)
  end)
```

This still wouldn't work because there's no way to force the spawned window to be only on the main screen and `gui_window:set_position(x, y)` appears to be ignored if `{ position = { x = 0, y = 0, origin = "MainScreen" } }` is provided on the `spawn_window` line.

**Reply by halostatue** [reply]
Note: with nightly at least as early as 20241015 and continuing through 20241101, I have observed the following issues:

- `position.origin` is ignored. I never want new windows opened on my secondary screen but wezterm will launch on the screen which has the active application when launched from `WezTerm.app`.
- `position.x` appears to be interpreted differently than 20240203. I have to change it to `main_screen.width + 144` to have the window launch on the right side of the window (which doesn't work in any case because `position.origin` is ignored).
- `wezterm.on('gui-attached', …)` is causing some sort of of infinite loop / hang when I try to set it up even to maximize the window on launch.

@wez should I open tickets for the above? I don't really have steps for replication.

**Reply by wez** [reply]
Sure, please file an issue for each distinct issue. Note that without a reproduction scenario it will make it harder to do anything about it.

**Reply by junguler** [reply]
thanks for this, fixed the last issue i had and helped me set wezterm exactly as i wanted

**Comment by Alien-lv426** [comment]
You may want to look at wezterm.on('gui-attached') rather than 'gui-startup', which is what I'm using to decide where to place the initial window
